DOC-2159: document four user-formatting control options (#2851)
tinymce/tinymce-docs: modules/ROOT/partials/configuration/text_color.adoc

});
----

[[color_cols_background]]
=== `color_cols_background`

This option specifies the number of columns for background text color grids.

> If `color_cols_background` is not set, the number of columns a background text color grid takes is set by the xref:#color_cols[`color_cols`] option.

Reviewer: ```suggestion
If `color_cols_background` is not set, the number of columns a highlight color grid takes is set by the xref:#color_cols[`color_cols`] option.
```
I think it's usually referred to as highlight color

If both `color_cols` and `color_cols_background` options are set, the `color_cols_background` value is used.

> If neither `color_cols` nor `color_cols_background` options are set, background text color grids present with the default number of columns, **5**.

Reviewer: ```suggestion
If neither `color_cols` nor `color_cols_background` options are set, the highlight color grid presents with the default number of columns, **5**.
```

> *Type:* `+Integer+`

Reviewer: ```suggestion
*Type:* `+Number+`
```
Technically it's a number since we handle decimal values by rounding.

*Default value:* 5

==== Example: using `color_cols_background`

> This example sets the number of text color grid columns to **10** then sets the background text color grid to display **3** columns.

Reviewer: ```suggestion
This example sets the number of columns in the default color grid to **10** then sets the background text color grid to display **3** columns.
```
color_cols also affects the color grids in the table dialog

[source,js]
----
tinymce.init({
  selector: 'textarea',  // change this value according to your html
  toolbar: 'backcolor forecolor',
  color_cols: 10,
  color_cols_background: 3,
});
----

[[color_cols_foreground]]
=== `color_cols_foreground`

This option specifies the number of columns for foreground text color grids.

> If `color_cols_foreground` is not set, the number of columns a background text color grid takes is set by the xref:#color_cols[`color_cols`] option.

Reviewer: ```suggestion
If `color_cols_foreground` is not set, the number of columns a foreground text color grid takes is set by the xref:#color_cols[`color_cols`] option.
```

If both `color_cols` and `color_cols_foreground` options are set, the `color_cols_foreground` value is used.

> If neither `color_cols` nor `color_cols_foreground` options are set, background text color grids present with the default number of columns, **5**.

Reviewer: ```suggestion
If neither `color_cols` nor `color_cols_foreground` options are set, text color grid presents with the default number of columns, **5**.
```

*Type:* `+Integer+`

*Default value:* 5

==== Example: using `color_cols_foreground`

> This example sets the number of text color grid columns to **10** then sets the foreground text color grid to display **3** columns.

Reviewer: ```suggestion
This example sets the number of columns in the default color grid to **10** then sets the foreground text color grid to display **3** columns.
```

[source,js]
----
tinymce.init({
  selector: 'textarea',  // change this value according to your html
  toolbar: 'backcolor forecolor',
  color_cols: 10,

>   color_cols_foreground: 3

Reviewer 2: ```suggestion
  color_cols_foreground: 3,
```

});
----

[[color_map]]
=== `+color_map+`

> This option allows for specifying a map of the text colors that will appear in the grid.

Reviewer: ```suggestion
This option allows for specifying a map of the text colors that will appear in the grid.

Colors are added to the array with a Hex or RGB color value, followed by the name of that color as it would appear in the color grid when hovered over. Both values are required for each color added to the color map.
```
Since I don't think this is described anywhere. Sounds like a weird method to me but that's what it looks like it expects.

This option allows for specifying a map of the text colors that will appear in the background text color grid.

If it is not set, the background text color grid takes it values from the `+color_map+` array.

And, if the `+color_map+` array is, further, not set, the background text color grid takes it values from the {productname} default color set.

That is, the background text color grid takes its values in the following priority order:

. a set `color_map_background` color array is used in preference to
. a set `color_map` color array, which is used in preference to
. the default {productname} color array.

*Type:* `+Array+`

==== Example: using `color_map_background`

[source,js]
----
tinymce.init({
  selector: 'textarea',  // change this value according to your html
  toolbar: 'forecolor backcolor',
  color_map_background: [
    '000000', 'Black',
    '808080', 'Gray',
    'FFFFFF', 'White',
    'FF0000', 'Red',
    'FFFF00', 'Yellow',
    '008000', 'Green',
    '0000FF', 'Blue'
  ]
});
----

When end-users add a further custom color to the background text color grid, that color is added to the background text color grid and stored in the `color_map_background` palette.

> NOTE: When a user adds a custom color to the `color_map_background` palette, the editor does not replicate the new color in the `color_map_foreground` palette.

Reviewer: Have you tested this? Seems like something only in local storage, not within the option value. It might appear in both if a user selects custom colors for later use.

Author: To quote my commit message:

> After further testing, re-wrote notes about end-user customisation and placed re-write into general IMPORTANT admonition set after all three color map array sections.

});
----

> If `+color_map+` is not explicitly set, {productname} populates the foreground and background text color grids with a default set of 22 colors.

Reviewer: ```suggestion
If `+color_map+` is not explicitly set, {productname} populates color grids with a default set of 22 colors.
```
It also affects the color grids in the table dialog

==== Example: explicitly setting `+color_map+` to the default {productname} color set

[source,js]
----
tinymce.init({
  selector: 'textarea',  // change this value according to your HTML
  toolbar: 'forecolor backcolor',
  color_map: [
    '#BFEDD2', 'Light Green',
    '#FBEEB8', 'Light Yellow',
    '#F8CAC6', 'Light Red',
    '#ECCAFA', 'Light Purple',
    '#C2E0F4', 'Light Blue',
    '#2DC26B', 'Green',
    '#F1C40F', 'Yellow',
    '#E03E2D', 'Red',
    '#B96AD9', 'Purple',
    '#3598DB', 'Blue',
    '#169179', 'Dark Turquoise',
    '#E67E23', 'Orange',
    '#BA372A', 'Dark Red',
    '#843FA1', 'Dark Purple',
    '#236FA1', 'Dark Blue',
    '#ECF0F1', 'Light Gray',
    '#CED4D9', 'Medium Gray',
    '#95A5A6', 'Gray',
    '#7E8C8D', 'Dark Gray',
    '#34495E', 'Navy Blue',
    '#000000', 'Black',
    '#ffffff', 'White'
  ]
});
----

[[color_map_background]]
=== `color_map_background`

> This option allows for specifying a map of the text colors that will appear in the background text color grid.

Reviewer: ```suggestion
This option allows for specifying a map of the text colors that will appear in the highlight color grid.
```

> If it is not set, the background text color grid takes it values from the `+color_map+` array.

Reviewer: ```suggestion
If it is not set, the highlight color grid takes it values from the `+color_map+` array.
```

> And, if the `+color_map+` array is, further, not set, the background text color grid takes it values from the {productname} default color set.
> 
> That is, the background text color grid takes its values in the following priority order:
> 
> . a set `color_map_background` color array is used in preference to
> . a set `color_map` color array, which is used in preference to
> . the default {productname} color array.

Reviewer: ```suggestion
And, if the `+color_map+` array is, further, not set, the highlight color grid takes it values from the {productname} default color set.

That is, the highlight grid takes its values in the following priority order:

. a set `color_map_background` color array is used in preference to
. a set `color_map` color array, which is used in preference to
. the default {productname} color array.
```

*Type:* `+Array+`

==== Example: using `color_map_background`

[source,js]
----
tinymce.init({
  selector: 'textarea',  // change this value according to your html
  toolbar: 'forecolor backcolor',
  color_map_background: [
    '000000', 'Black',
    '808080', 'Gray',
    'FFFFFF', 'White',
    'FF0000', 'Red',
    'FFFF00', 'Yellow',
    '008000', 'Green',
    '0000FF', 'Blue'
  ]
});
----

[[color_map_foreground]]
=== `color_map_foreground`

> This option allows for specifying a map of the text colors that will appear in the foreground text color grid.
> 
> If it is not set, the foreground text color grid takes it values from the `+color_map+` array.
> 
> And, if the `+color_map+` array is, further, not set, the background text color grid takes it values from the {productname} default color set.
> 
> That is, the foreground text color grid takes its values in the following priority order:
> 
> . a set `color_map_foreground` color array is used in preference to
> . a set `color_map` color array, which is used in preference to
> . the default {productname} color array.

Reviewer: ```suggestion
This option allows for specifying a list of the text colors that will appear in the text color grid.

If it is not set, the text color grid takes it values from the set `+color_map+` array.

And, if the `+color_map+` array is, further, not set, the text color grid takes it values from the {productname} default color set.

That is, the text color grid takes its values in the following priority order:

. a set `color_map_foreground` color array is used in preference to
. a set `color_map` color array, which is used in preference to
. the default {productname} color array.
```
